gax/src/test/java/com/google/api/gax/rpc/ClientContextTest.java

```diff
@@ -769,4 +786,76 @@ public void testExecutorSettings() throws Exception {
     transportChannel = (FakeTransportChannel) context.getTransportChannel();
     assertThat(transportChannel.getExecutor()).isSameInstanceAs(executorProvider.getExecutor());
   }
+
+  @Test
+  public void testAddApiKeyToHeadersFromStubSettings() throws IOException {
+    StubSettings settings = new FakeStubSettings.Builder().setApiKey("stub-setting-key").build();
+    EnvironmentProvider environmentProvider = new FakeEnvironmentProvider("env-key", null);
+    Map<String, String> headers = new HashMap<>();
+    ClientContext.addApiKeyToHeaders(settings, environmentProvider, headers);
+    assertTrue(headers.containsKey("x-goog-api-key"));
+    assertEquals("stub-setting-key", headers.get("x-goog-api-key"));
+  }
+
+  @Test
+  public void testAddApiKeyToHeadersFromEnvironmentProvider() throws IOException {
+    StubSettings settings = new FakeStubSettings.Builder().build();
+    EnvironmentProvider environmentProvider = new FakeEnvironmentProvider("env-key", null);
+    Map<String, String> headers = new HashMap<>();
+    ClientContext.addApiKeyToHeaders(settings, environmentProvider, headers);
+    assertTrue(headers.containsKey("x-goog-api-key"));
+    assertEquals("env-key", headers.get("x-goog-api-key"));
+  }
+
+  @Test
+  public void testAddApiKeyToHeadersNoApiKey() throws IOException {
+    StubSettings settings = new FakeStubSettings.Builder().build();
+    EnvironmentProvider environmentProvider = new FakeEnvironmentProvider(null, null);
+    Map<String, String> headers = new HashMap<>();
+    ClientContext.addApiKeyToHeaders(settings, environmentProvider, headers);
+    assertFalse(headers.containsKey("x-goog-api-key"));
+  }
+
+  @Test
+  public void testAddApiKeyToHeadersThrows() throws IOException {
+    StubSettings settings = new FakeStubSettings.Builder().build();
+    EnvironmentProvider environmentProvider =
+        new FakeEnvironmentProvider("env-key", "/path/to/adc/json");
+    Map<String, String> headers = new HashMap<>();
+    try {
+      ClientContext.addApiKeyToHeaders(settings, environmentProvider, headers);
+      fail("should throw an exception");
+    } catch (IOException e) {
+      assertTrue(
+          "expected retrieveApiKey to throw an exception",
+          e.getMessage()
+              .contains(
+                  "Environment variables GOOGLE_API_KEY and GOOGLE_APPLICATION_CREDENTIALS are mutually exclusive"));
+    }
+  }
+
+  @Test
+  public void testApiKey() throws IOException {
+    FakeStubSettings.Builder builder = new FakeStubSettings.Builder();
+
+    FakeTransportChannel transportChannel = FakeTransportChannel.create(new FakeChannel());
+    FakeTransportProvider transportProvider =
+        new FakeTransportProvider(transportChannel, null, true, null, null);
+    builder.setTransportChannelProvider(transportProvider);
+
+    HeaderProvider headerProvider = Mockito.mock(HeaderProvider.class);
+    Mockito.when(headerProvider.getHeaders()).thenReturn(ImmutableMap.<String, String>of());
```

Comment:
Can it be
```suggestion
    Mockito.when(headerProvider.getHeaders()).thenReturn(ImmutableMap.<>of());
```
(or just ImmutableMap.of() without a warning)?